### PR TITLE
Update appmodule to use DSL

### PR DESCRIPTION
**Pull request by ranchalp**

### CODE REVIEW

**Comment by ranchalp** [comment]
Related with this:

Why not have all default modules inside `modulesWithDefaults` directly? i.e.:
```
	// Factory module with instances of the checkpointing protocol.
    checkpointing := checkpoint.Factory(common.DefaultModuleConfig(), ownID, logging.Decorate(logger, "CHKP: "))


    // PBFT module with instances of the pbft protocol as segments to be called by ISS.
    ordering := orderers.Factory(orderers.DefaultModuleConfig(), params.Iss, ownID, hashImpl, cryptoImpl, …	
…
    batchFetcher := batchfetcher.NewModule(
   	…
    )
```

File: pkg/trantor/appmodule.go [hunk 1]
```diff
@@ -24,121 +28,80 @@ type AppModule struct {
...
-func (m *AppModule) ImplementsModule() {}
+	// UponNewOrderedBatch sequentially apply a batch of ordered transactions to the application logic
+	// and returns an empty event list.
+	bfpbdsl.UponNewOrderedBatch(m, func(txs []*trantorpbtypes.Transaction) error {
 
-// applyNewOrderedBatch sequentially applies a batch of ordered transactions to the application logic
-// and returns an empty event list.
-func (m *AppModule) applyNewOrderedBatch(batch *bfpb.NewOrderedBatch) (*events.EventList, error) {
+		if err := appModule.appLogic.ApplyTXs(
+			sliceutil.Transform(
```
Comment (matejpavlovic) [inline pkg/trantor/appmodule.go]: Change the AppLogic interface to accept the generated type.

**Review by matejpavlovic — CHANGES_REQUESTED** [review]
Thanks Alex!
Just minor comments.

**Comment by matejpavlovic** [comment]
Regarding the default modules, it's a separate issue (I've just created one on GitHub), see #431

File: pkg/trantor/testutils.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"github.com/filecoin-project/mir/pkg/eventmangler"
+	"github.com/filecoin-project/mir/pkg/iss"
+)
+
+// PerturbMessages configures the SMR system to randomly drop and delay some of the messages sent over the network.
+// Useful for debugging and stress-testing.
+// The params argument defines parameters of the perturbation, such as how many messages should be dropped
+// and how the remaining messages should be delayed.
+func PerturbMessages(params *eventmangler.ModuleParams, sys *System) error {
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/trantor/testutils.go:12]: done

File: pkg/trantor/appmodule.go [hunk 3]
```diff
@@ -31,7 +31,7 @@ type AppModule struct {
 }
 
 // NewAppModule creates a new AppModule.
-func NewAppModule(appLogic AppLogic, transport net.Transport, protocolModule t.ModuleID) modules.PassiveModule {
+func NewAppModule(appLogic AppLogic, transport net.Transport, _ t.ModuleID) modules.PassiveModule {
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/trantor/appmodule.go]: done

File: pkg/trantor/appmodule.go [hunk 4]
```diff
@@ -24,121 +28,80 @@ type AppModule struct {
...
-	default:
-		return nil, fmt.Errorf("unexpected type of App event: %T", event.Type)
+	appModule := &AppModule{
+		appLogic:  appLogic,
+		transport: transport,
 	}
 
-	return events.EmptyList(), nil
-}
+	m := dsl.NewModule("appModule")
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/trantor/appmodule.go]: done

File: pkg/trantor/appmodule.go (hunk 1 above)
Comment (ranchalp) [reply pkg/trantor/appmodule.go]: done

File: pkg/trantor/appmodule.go [hunk 5]
```diff
@@ -24,121 +28,80 @@ type AppModule struct {
...
-	}
-	m.transport.Connect(membership)
-	// TODO: Save the origin module ID in the event and use it here, instead of saving the m.protocolModule.
-	return events.ListOf(isspbevents.NewConfig(
-		m.protocolModule,
-		newEpoch.EpochNr,
-		membership,
-	).Pb()), nil
-}
+	// UponRestoreState restore the application state from a snapshot.
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/trantor/appmodule.go]: done

File: pkg/trantor/appmodule.go [hunk 6]
```diff
@@ -24,121 +28,80 @@ type AppModule struct {
...
+
+	// UponNewEpoch inform the application logic of the new epoch and emit an event (to the protocol module)
+	// containing the configuration for the new epoch.
+	apppbdsl.UponNewEpoch(m, func(epochNr types.EpochNr, protocolModule t.ModuleID) error {
+		membership, err := appModule.appLogic.NewEpoch(epochNr)
+		if err != nil {
+			return fmt.Errorf("error handling NewEpoch event: %w", err)
+		}
+		appModule.transport.Connect(membership)
+		// TODO: Save the origin module ID in the event and use it here, instead of saving the m.protocolModule.
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/trantor/appmodule.go]: Yes, I missed one last commit & push but this was solved locally. Just pushed with latest commit that resolves all comments.

[the remark this answers is not in the document]
Comment (matejpavlovic) [reply pkg/trantor/appmodule.go]: Still there 😄

**Review by matejpavlovic — APPROVED** [review]
LGTM, feel free to resolve the last one directly after fixing it and merge. Thanks!

**Pull request merged**